### ISSUE TITLE
Karaf 4.0.8 should be tested

### ISSUE THREAD

**Issue by maggu2810** [openhab/openhab-distro#18]
``` text
Jean-Baptiste Onofré, 08:48
...
Today I'm fixing the latest Jira for 4.0.4, I should be able to submit 4.0.4 release to vote tomorrow
```

We should test K404 as soon as its voting round starts and we should vote +1 (if all is fine) or -1 (is something is broked) for the release.
##

**Comment by maggu2810** [comment]
FYI: the vote has been started
http://karaf.922171.n3.nabble.com/VOTE-Apache-Karaf-4-0-4-release-tp4044726.html
After the release freeze we should test this immediately to guarantee that we have no regressions using K404 for the next developmemt round.

**Comment by maggu2810** [comment]
K404 has been released:
http://karaf.922171.n3.nabble.com/ANN-Apache-Karaf-4-0-4-released-tp4044829.html

**Comment by kaikreuzer** [comment]
I just had a first try (see https://github.com/kaikreuzer/openhab-distro/tree/karaf404)
Build works fine, but the offline-distro does not start:
```
2016-01-16 01:07:00.038 [WARN ] [url.mvn.internal.AetherBasedResolver] - Error resolving artifactorg.ops4j.pax.web:pax-web-jetty:jar:4.2.4:Could not find artifact org.ops4j.pax.web:pax-web-jetty:jar:4.2.4
shaded.org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact org.ops4j.pax.web:pax-web-jetty:jar:4.2.4
        at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolve(DefaultArtifactResolver.java:444)[4:org.ops4j.pax.url.mvn:2.4.5]
        at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifacts(DefaultArtifactResolver.java:246)[4:org.ops4j.pax.url.mvn:2.4.5]
        at shaded.org.eclipse.aether.internal.impl.DefaultArtifactResolver.resolveArtifact(DefaultArtifactResolver.java:223)[4:org.ops4j.pax.url.mvn:2.4.5]
        at shaded.org.eclipse.aether.internal.impl.DefaultRepositorySystem.resolveArtifact(DefaultRepositorySystem.java:294)[4:org.ops4j.pax.url.mvn:2.4.5]
```
It seems as if we need to update some other versions with it.
@dvanherbergen Can you help on that?

**Comment by dvanherbergen** [comment]
We'll need to add a patched version of  pax-web-jetty 4.2.4. I will have a look later today.

**Comment by kaikreuzer** [comment]
> We'll need to add a patched version of pax-web-jetty 4.2.4. I will have a look later today.

Did you prepare such a patched version?

**Comment by dvanherbergen** [comment]
Yes, but I was having some issues getting it deployed correctly to bintray. Do you happen to have the correct command to use?

**Comment by kaikreuzer** [comment]
You mean
```
curl -T <FILE.EXT> -u<USER>:<API_KEY> https://api.bintray.com/content/openhab/mvn/openhab-distro-deps/<VERSION>/<FILE_TARGET_PATH>
```
or better directly through Maven deploy with
```
<distributionManagement>
    <repository>
        <id>bintray-openhab-mvn</id>
        <name>openhab-mvn</name>
        <url>https://api.bintray.com/maven/openhab/mvn/openhab-distro-deps/;publish=1</url>
    </repository>
</distributionManagement>
```
?

**Comment by dvanherbergen** [comment]
It's the maven deploy that was causing trouble for me. I got it to work now after throwing out the nexus-staging-maven-plugin, which was interfering with a regular deploy.

The jar is available now on bintray as
```
	<groupId>org.openhab.distro.deps</groupId>
	<artifactId>pax-web-jetty</artifactId>
	<version>4.2.4.sp1</version>
```

**Comment by dvanherbergen** [comment]
I will do a test with an updated pom + descriptor. If the offline distro starts, I'll create a PR.

**Comment by maggu2810** [comment]
Apr 01, 2016; 3:23am: [VOTE] Apache Karaf 4.0.5 release
http://karaf.922171.n3.nabble.com/VOTE-Apache-Karaf-4-0-5-release-td4046085.html

**Comment by kaikreuzer** [comment]
I have updated the title accordingly - it imho makes no sense to go to 4.0.4 anymore.

**Comment by maggu2810** [comment]
> The jar is available now on bintray as
```
	<groupId>org.openhab.distro.deps</groupId>
	<artifactId>pax-web-jetty</artifactId>
	<version>4.2.4.sp1</version>
```

@dvanherbergen Cannot find it. Could you point me to?

**Comment by dvanherbergen** [comment]
oops, my bad. The version is slightly different. You can find it [here] (https://bintray.com/openhab/mvn/openhab-distro-deps/4.2.4/view#files/org/openhab/distro/deps/pax-web-jetty/4.2.4)
```
<dependency>
  <groupId>org.openhab.distro.deps</groupId>
  <artifactId>pax-web-jetty</artifactId>
  <version>4.2.4</version>
</dependency>
```

**Comment by maggu2810** [comment]
@dvanherbergen Thanks!
Do you know the issue about the console (keyboard input corrupt, double openHAB logo)? I don't find it ATM.

**Comment by dvanherbergen** [comment]
Yes, it happens when you start using the start.sh script. Then the logo appears twice (with version 4.0.4) only. It has been some time since I tested, this so I'm not sure if it is still there. Potentially this could have been resolved if one of the dependencies has been changed.

**Comment by maggu2810** [comment]
I used the openhab stuff yesterday to create some PRs.
After I tested my changes, I see this error myself.
I bumped Karaf from 4.0.3 to 4.0.4 and the error has gone.
Are you sure you have seen this with Karaf 4.0.4 and not with Karaf 4.0.3?
IIRC there has been a separate issue for this problem (but perhaps I am wrong, did not found them) and I want to report there that it could be solved after using a more recent Karaf (hopefully 4.0.5 soon).

**Comment by dvanherbergen** [comment]
I believe it was 4.0.4, but I will retest to confirm.

**Comment by kaikreuzer** [comment]
I updated the title to 4.0.6, which had just been released.

**Comment by kaikreuzer** [comment]
@dvanherbergen For Karaf 4.0.6, we now need a patched org.ops4j.pax.web:pax-web-jetty:jar:4.2.8. It is a bit awkward having to ask you every time for a patched version; but I also do not have the details on how you patched it...
Could you come up with an easy script that does the necessary changes, so that we could more or less do this automatically? Or what is the best way to proceed here for easy Karaf upgrades?

**Comment by dvanherbergen** [comment]
I'll have a look...

**Comment by maggu2810** [comment]
No need to test K406 and K407 if you need a working shell.

**Comment by kaikreuzer** [comment]
Why? Could you link the issue?

**Comment by maggu2810** [comment]
https://issues.apache.org/jira/browse/KARAF-4714

**Comment by kaikreuzer** [comment]
:-( Seems we are the only ones using it on ARM...
But I would hope that this issue gets addressed before 4.0.7 is final.

**Comment by dvanherbergen** [comment]
> Could you come up with an easy script that does the necessary changes, so that we could more or less do this automatically? Or what is the best way to proceed here for easy Karaf upgrades?

I've finally found some time to look into this. Pull request #285  automates the patch as part of the build. It should work with any version of pax-web as long as the method signature of org.ops4j.pax.web.utils.ClassPathUtil.findResources doesn't change. Looking at the history of that file, it hasn't changed much in the last years, so I think that should be OK.

The patch automation itself is done using a small utility jar that performs byte code manipulation. You can find that jar here: https://github.com/dvanherbergen/openhab-util. I've also deployed it to bintray so that it can be used in the build.

**Comment by kaikreuzer** [comment]
Hi Davy, This is pretty cool, many thanks! This should make it much easier for us in future to do Karaf upgrades. We now just have to find a new Karaf version that hasn't any major flaws ;-)

**Comment by maggu2810** [comment]
Interesting approach. :+1:

**Comment by kaikreuzer** [comment]
FTR: https://issues.apache.org/jira/browse/KARAF-4714 is only fixed in 4.0.8, so we have to wait for this to be released.
@maggu2810 Did you by any chance already verify that 4.0.8 correctly works again?

**Comment by maggu2810** [comment]
It is not solved for me.

**Comment by kaikreuzer** [comment]
:-(
But many thanks for testing!

**Comment by maggu2810** [comment]
I will keep you informed if the console problem is solved in general (at least on my ARM systems).

**Comment by kaikreuzer** [comment]
FTR: I am working on the Karf 4.0.8 upgrade and it looks quite good so far. A few small issues still to solve, though.

**Comment by maggu2810** [comment]
> A few small issues still to solve, though.

Can you give me some details?

**Comment by maggu2810** [comment]
Do you refer to openhab/openhab-core#99

**Comment by kaikreuzer** [comment]
Actually only to [this single line](https://github.com/openhab/openhab-core/pull/99/files#diff-600376dffeb79835ede4a0b285078036R62). I hadn't changed this in my first tests and that resulted in having two Karaf shells being active in parallel - not a nice experience ;-

The rest works pretty smooth!
